File: src/node/magma/app/compile/type/TemplateType.ts
import { Type } from "../../../../magma/api/Type";
import { List } from "../../../../magma/api/collect/list/List";
import { Main } from "../../../../magma/app/Main";
import { Strings } from "../../../../jvm/api/text/Strings";
export class TemplateType implements Type {
	base: string;
	args: List<string>;
	constructor (base: string, args: List<string>) {
		this.base = base;
		this.args = args;
	}
	static generateValueStrings(values: List<string>): string {
		return Main/*auto*/.generateAll(values/*auto*/, TemplateType/*auto*/.mergeValues);
	}
	static mergeValues(cache: string, element: string): string {
		if (Strings/*auto*/.isEmpty(cache/*auto*/)) {
			return cache/*auto*/ + element/*auto*/;
		}
		return cache/*auto*/ + ", " + element/*auto*/;
	}
	generate(): string {
		return this/*auto*/.base + "<" + TemplateType/*auto*/.generateValueStrings(this/*auto*/.args) + ">";
	}
	isFunctional(): boolean {
		return false/*auto*/;
	}
	isVar(): boolean {
		return false/*auto*/;
	}
	generateBeforeName(): string {
		return "";
	}
}

File: src/java/magma/app/compile/type/TemplateType.java
package magma.app.compile.type;

import magma.api.collect.list.Iterable;
import magma.app.compile.merge.Merger;
import magma.app.compile.merge.ValueMerger;

public record TemplateType(String base, Iterable<String> args) implements Type {
    @Override
    public String generate() {
        return this.base + "<" + Merger.generateAll(this.args, new ValueMerger()) + ">";
    }

    @Override
    public boolean isFunctional() {
        return false;
    }

    @Override
    public boolean isVar() {
        return false;
    }

    @Override
    public String generateBeforeName() {
        return "";
    }

    @Override
    public String generateSimple() {
        return this.base;
    }
}
